### PR TITLE
event: delayed handling of window property changes

### PR DESCRIPTION
**Pull request by yshui**
Move handling window property changes into the X critical section.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

### CODE REVIEW

File: src/win.c [hunk 1]
```diff
@@ -2374,6 +2453,44 @@ void win_clear_flags(struct managed_win *w, uint64_t flags) {
...
+	const auto bits_per_element = sizeof(*w->stale_props) * 8;
+	if (prop >= w->stale_props_capacity * bits_per_element) {
+		const auto new_size = prop / bits_per_element + 1;
+		w->stale_props = realloc(w->stale_props, new_size * sizeof(*w->stale_props));
+
+		// Clear the content of the newly allocated bytes
+		memset(w->stale_props + w->stale_props_capacity, 0,
+		       new_size - w->stale_props_capacity);
+		w->stale_props_capacity = new_size;
+	}
```
Comment (tryone144) [inline src/win.c:2484]: Is it okay to loose all previously set property flags, when we have to expand (i.e. re-allocate) our bitmask?

File: src/win.c [hunk 2]
```diff
@@ -320,11 +345,81 @@ void win_release_images(struct backend_base *backend, struct managed_win *w) {
...
 		map_win_start(ps, w);
 		win_clear_flags(w, WIN_FLAGS_MAPPED);
 	}
+
+	// Check client first, because later property updates need accurate client window
+	// information
+	if (win_check_flags_all(w, WIN_FLAGS_CLIENT_STALE)) {
+		win_recheck_client(ps, w);
+		win_clear_flags(w, WIN_FLAGS_CLIENT_STALE);
+	}
```
Comment (tryone144) [inline src/win.c:364]: Doing this check/update here should fix a "rare" race-condition when a client window disappears and the frame is still rendered in the next cycle, but it's shadow has been invalidated due to specific shadow-rules (see #479). Should fix the crashes described in #390 and #487.

File: src/win.c (hunk 1 above)
Comment (yshui) [reply src/win.c:2484]: No, the memset only clear the newly allocated bytes.

Comment (tryone144) [reply src/win.c:2484]: Sorry, my bad. `realloc()` does in fact preserve its content. Had that mistaken with another function.

Comment (yshui) [reply src/win.c:2484]: Ah, wait. I forgot to `* sizeof(*w->stale_props)` in the memset.

File: src/win.c (hunk 2 above)
Comment (yshui) [reply src/win.c:364]: OK, can I merge this first, and you rebase #479 ? Or do you want that to be merged first?

Comment (tryone144) [reply src/win.c:364]: Merge this first. I'm gona rebase #479 with the mentioned comments. Doing the check/update here seem more plausible.

**Pull request merged**